src/AudioSocket.cpp

```diff
@@ -0,0 +1,665 @@
+        mNumChannels = nChansIn;
+    } else {
+        mNumChannels = nChansOut;
+    }
+}
+
+//*******************************************************************************
+void FromAudioSocketPlugin::gotAudioHeader(int samplingRate, int bufferSize)
+{
+    mRemoteSampleRate      = samplingRate;
+    mRemoteBufferSize      = bufferSize;
+    mRemoteBytesPerChannel = bufferSize * sizeof(float);
+    mNextExtraSample       = 0;
+    mLastExtraSample       = 0;
+    mQueueCheckSec         = 2;
+    mRemoteIsReady         = true;
+    resetQueueStats();
+}
+
+//*******************************************************************************
+void FromAudioSocketPlugin::gotConnection()
+{
+    mRemoteSampleRate      = 0;
+    mRemoteBufferSize      = 0;
+    mRemoteBytesPerChannel = 0;
+    mRemoteIsReady         = false;
+    mIsConnected           = true;
+}
+
+//*******************************************************************************
+void FromAudioSocketPlugin::lostConnection()
+{
+    mIsConnected = false;
+}
+
+//*******************************************************************************
+void FromAudioSocketPlugin::resetQueueStats()
+{
+    mMinQueuePackets = AudioSocketMaxQueueSize;
+    mMaxQueuePackets = 0;
+    mNextQueueCheck  = getSampleRate() * mQueueCheckSec;
+    if (mQueueCheckSec < 512)  // max interval of about 8.5 minutes
+        mQueueCheckSec *= 2;
+}
+
+//*******************************************************************************
+AudioSocketWorker::AudioSocketWorker(AudioSocketQueueT& sendQueue,
+                                     AudioSocketQueueT& receiveQueue,
+                                     QSharedPointer<QLocalSocket>& s)
+    : mSendQueue(sendQueue), mReceiveQueue(receiveQueue), mSocketPtr(s)
+{
+    mSendBuffer.resize(AudioSocketMaxSamplesPerBlock * AudioSocketNumChannels
+                       * sizeof(float));
+    mRecvBuffer.resize(AudioSocketMaxSamplesPerBlock * AudioSocketNumChannels
+                       * sizeof(float));
+}
+
+//*******************************************************************************
+AudioSocketWorker::~AudioSocketWorker() {}
+
+//****************************************************************************
+void AudioSocketWorker::start()
+{
+    setRealtimeProcessPriority();
+}
+
+//****************************************************************************
+void AudioSocketWorker::connect()
+{
+    if (isConnected()) {
+        return;
+    }
+
+    SocketClient c(mSocketPtr);
+
+    if (!c.connect()) {
+        emit signalConnectionFailed();
+        return;
+    }
+
+    if (!c.sendHeader("audio")) {
+        mSocketPtr->close();
+        emit signalConnectionFailed();
+        return;
+    }
+
+    cout << "Established audio socket connection" << endl;
+    emit signalConnectionEstablished();
+}
+
+//*******************************************************************************
+void AudioSocketWorker::close()
+{
+    if (!isConnected()) {
```

Comment:
This seems like a very rare case but `mSocketPtr->state()` could be `QLocalSocket::ConnectingState` as it's initiating the connection. If `AudioSocketWorker::close()` is called during this transitory phase, the socket wouldn't get closed and might leave a dangling connection behind?

Reply:
I don't think it's possible because we block on connecting and only the worker thread can perform the socket operations, but I added an extra check to be safe here https://github.com/jacktrip/jacktrip/pull/1368/commits/091c5ff02b14f630fc2c888c99736547af7fe228